loom/engine/bindings/loom/lmCSS.cpp

```diff
@@ -0,0 +1,1178 @@
+                    attribute.assign(start, p - start);
+                    style->addAttribute(attribute);
+
+                    if (p[0] == ' ')
+                    {
+                        p = skipCommentsAndWS(p, end);
+                        start = p;
+                    }
+
+                    if (p[0] == ',')
+                    {
+                        p++;
+                        start = p;
+                        continue;
+                    }
+
+                    if (p[0] == ']')
+                    {
+                        break;
+                    }
+                }
+
+                p++;
+            }
+
+            // skip ]
+            p++;
+
+            return p;
+        }
+
+        static const char* readKey(const char* p, const char* end, utString& key)
+        {
+            const char* start = p;
+            while (p < end && p[0] != ' ' && p[0] != ':')
+            {
+                p++;
+            }
+
+            key.assign(start, p - start);
+
+            return p;
+        }
+
+        static const char* parseValueString(const char* p, const char* end, utString& value, char quote)
+        {
+            // skip quote
+            p++;
+
+            const char* start = p;
+            while (p < end && p[0] != quote && p[0] != '\n')
+            {
+                p++;
+            }
+
+            value.assign(start, p - start);
+
+            if (p >= end)
+            {
+                lmLogError(cssLogGroup, "Syntax error: Unexpected EOF");
+                return p;
+            }
+
+            if (p[0] == '\n')
+            {
+                lmLogError(cssLogGroup, "Syntax error: Unexpected newline, unterminated string");
+                return p;
+            }
+
+            if (p[0] == quote)
+            {
+                // skip quote
+                p++;
+            }
+
+            return p;
+        }
+
+        static const char* parseValue(const char* p, const char* end, utString& value)
+        {
+            const char* start = p;
+
+            while (p < end && p[0] != ';' && p[0] != '\n')
+            {
+                p++;
+            }
+
+            value.assign(start, p - start);
+
+            if (p >= end)
+            {
+                lmLogError(cssLogGroup, "Syntax error: Unexpected EOF");
+                return p;
+            }
+
+            if (p[0] == '\n')
+            {
+                lmLogError(cssLogGroup, "Syntax error: Unexpected newline, expected ';'");
+            }
+
+            return p;
+        }
+
+        static const char* parseProperty(Style* style, const char* p, const char* end)
+        {
+            utString key;
+            utString value;
+
+            p = skipCommentsAndWS(p, end);
+            p = readKey(p, end, key);
+            p = skipCommentsAndWS(p, end);
+
+            if (p >= end)
+            {
+                lmLogError(cssLogGroup, "Syntax error: Unexpected EOF");
```

Comment:
All css errors should include a source line/file number so that devs don't hate us. (even if we have to set source file on the parser explicitly).